### PR TITLE
Query and Update Animals Table

### PR DESCRIPTION
**Pull request by momo-87**
### Added Features:
- Add a column species of type string to animals table
- Insert new data into animals table
- Set the species column to unspecified inside a transaction then rollback
- Update the animals table by setting the species column to 'digimon' for all animals that have a name ending in 'mon'
- Update the animals table by setting the species column to 'pokemon' for all animals that don't have species already set
- Commit the transaction
- In a transaction delete all records in the animals, then roll back the transaction
- Delete all animals born after Jan 1st, 2022
- Create a savepoint for the transaction
- Update all animals' weight to be their weight multiplied by -1
- Rollback to the savepoint
- Update all animals' weights that are negative to be their weight multiplied by -1
- Commit transaction
- Write a query to answer the question: How many animals are there?
- Write a query to answer the question: How many animals have never tried to escape?
- Write a query to answer the question: What is the average weight of animals?
- Write a query to answer the question: Who escapes the most, neutered or not neutered animals?
- Write a query to answer the question: What is the minimum and maximum weight of each type of animal?
- Write a query to answer the question: What is the average number of escape attempts per animal type of those born between 1990 and 2000?
screenshot1
screenshot2
screenshot3
screenshot4
screenshot5

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -22,6 +22,61 @@ select * from animals where name != 'Gabumon';
 -- Find all animals with a weight between 10.4kg and 17.3kg (including the animals with the weights that equals precisely 10.4kg or 17.3kg)
 select * from animals where weight_kg between 10.4 and 17.3;
 
+-- Set the species column to unspecified inside a transaction then roll back
```
Comment (topeogunleye) [inline queries.sql:25]: Hello, I'm here to review your code. You have done a partial job on this project.

One thing I noticed is that you only updated the animals table where the species column is null. However, the project requirements ask you to update the entire species column to unspecified. You should remove the WHERE clause from your UPDATE statement to update all rows in the table.

Another thing I noticed is that you did not include any SELECT statements to verify that your change was made and rolled back. The project requirements explicitly ask you to do this step, so you should add two SELECT statements after your UPDATE and ROLLBACK statements, respectively. You should also check the output of your SELECT statements to make sure that they match the expected results.

Here's an example of how you can write your code that meets all the project requirements:

```sql
BEGIN; -- start transaction

UPDATE animals
SET species = 'unspecified'; -- make change

SELECT species from animals; -- verify that change was made

ROLLBACK; -- undo change

SELECT species from animals; -- verify that change was undone

COMMIT; -- end transaction

File: queries.sql [hunk 2]
```diff
@@ -22,6 +22,61 @@ select * from animals where name != 'Gabumon';
...
+begin;
+-- Update the animals table by setting the species column to 'digimon' for all animals that have a name ending in 'mon'
+update animals
+set species = 'digimon' where name like '%mon';
 
+-- Update the animals table by setting the species column to 'pokemon' for all animals that don't have species already set.
+update animals
+set species = 'pokemon' where species is null;
+-- Commit the trabsaction
+commit;
```
Comment (topeogunleye) [inline queries.sql]: Hello, I'm here to review your code. You have done a good job on this project.

One thing I noticed is that you did not include any SELECT statements to verify that your change was made and persisted after the commit. You should add two SELECT statements before and after your COMMIT statement, respectively, to check the output of your species column.

Here's an example of how you can write your code that meets all the project requirements:

```sql
BEGIN; -- start transaction

-- Update the animals table by setting the species column to 'digimon' for all animals that have a name ending in 'mon'
UPDATE animals
SET species = 'digimon' WHERE name LIKE '%mon';

-- Update the animals table by setting the species column to 'pokemon' for all animals that don't have species already set
UPDATE animals
SET species = 'pokemon' WHERE species IS NULL;

-- Verify that change was made
SELECT species FROM animals;

-- Commit the transaction
COMMIT;

-- Verify that change persists after commit
SELECT species FROM animals;

File: queries.sql [hunk 3]
```diff
@@ -22,6 +22,61 @@ select * from animals where name != 'Gabumon';
...
 
+-- start a transaction 2
+begin;
+-- delete all records in the animals
+delete from animals;
+-- roll back the transaction
+rollback;
```
Comment (topeogunleye) [inline queries.sql]: Another thing I noticed is that you did not include any SELECT statements to verify that your change was made and rolled back. The project requirements explicitly ask for this step, so you should add two SELECT statements after your DELETE and ROLLBACK statements, respectively. You should also check the output of your SELECT statements to make sure that they match the expected results.

Here's an example of how you can write your code that meets all the project requirements:

```sql
BEGIN; -- start transaction

-- Delete all records in the animals table
DELETE FROM animals;

-- Verify that change was made
SELECT COUNT(*) FROM animals;

-- Roll back the transaction
ROLLBACK;

-- Verify that change was rolled back
SELECT COUNT(*) FROM animals;

File: queries.sql [hunk 4]
```diff
@@ -22,6 +22,61 @@ select * from animals where name != 'Gabumon';
...
+select avg(weight_kg) from animals;
+-- Who escapes the most, neutered or not neutered animals?
+select neutered, max(escape_attempts) from animals
+group by neutered;
+-- What is the minimum and maximum weight of each type of animal?
+select species, min(weight_kg), max(weight_kg) from animals
+group by species;
+-- What is the average number of escape attempts per animal type of those born between 1990 and 2000?
+select species, avg(escape_attempts) from animals where date_of_birth between '1990-01-01' and '2000-12-31'
+group by species;
```
Comment (topeogunleye) [inline queries.sql:93]: Another thing I noticed is that you have a syntax error in your WHERE clause. You placed the WHERE clause after the GROUP BY clause, which is not allowed in SQL. The WHERE clause should always come before the GROUP BY clause, as it filters the rows before they are grouped.

Here's an example of how you can write your code that meets all the project requirements:

```sql
-- Select species and average number of escape attempts from animals
SELECT species, AVG(escape_attempts) FROM animals
-- Filter rows by date of birth between 1990 and 2000
WHERE date_of_birth >= '1990/01/01' AND date_of_birth <= '2000/12/31'
-- Group rows by species
GROUP BY species;

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @momo-87, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

## Highlights 💪 
- You added a new column called species with a string type to the animals table in the schema.sql file.
- You inserted the required data into the animals table in the data.sql file.
- You used the ALTER TABLE statement to add a new column to an existing table.

- You used the INSERT INTO statement to insert multiple rows of data into a table.
- You used the VALUES clause to specify the values for each column in each row.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- I added the below because there were no changes of the README file in your query-and-update-animals.
- I kindly suggest that you add the appropriate steps for running your project, deployment and running tests because the current tests you have in your README is for a React Project.

image


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql (hunk 1 above)
Comment (momo-87) [reply queries.sql:25]: Hi @topeogunleye , I really appreciate the time and effort you invested in reviewing my project. I set the species to 'unspecified' for all rows and I added the `select` statement where required for the first transaction.

Thank you for your feedback.

File: queries.sql (hunk 2 above)
Comment (momo-87) [reply queries.sql]: Hi @topeogunleye , I really appreciate the time and effort you invested in reviewing my project.  The `select` statement is shown in the first screenshot in the PR description. Anyway, I also added it where required for the second transaction.

Thank you for your feedback.

File: queries.sql (hunk 3 above)
Comment (momo-87) [reply queries.sql]: Hi @topeogunleye , I really appreciate the time and effort you invested in reviewing my project.  The `select` statements are used in the second screenshot in the PR description. Anyway, I also added `select count` statement where required for the third transaction.

Thank you for your feedback.

File: queries.sql (hunk 4 above)
Comment (momo-87) [reply queries.sql:93]: Hi @topeogunleye , I really appreciate the time and effort you invested in reviewing my project.  in the screenshot that you provided, the `where` clause is on line 81 while `group by` is on line 82. So I guess there is not any error in the code.

Thank you for your feedback.

**Review by alexander16108 — APPROVED** [review]
Hi @momo-87 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- You added a new column called species with a string type to the animals table in the schema.sql file.
- 
- You inserted the required data into the animals table in the data.sql file.
- 
- You used the ALTER TABLE statement to add a new column to an existing table.
- 
- You used the INSERT INTO statement to insert multiple rows of data into a table.
- 
- You used the VALUES clause to specify the values for each column in each row.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql (hunk 1 above)
Comment (topeogunleye) [reply queries.sql:25]: Hi, @momo-87,

You’re very welcome. I’m glad to hear that you made the necessary changes to your project. You have done a great job of following the instructions and applying the SQL concepts. Your code is well-written, clear, and efficient.

File: queries.sql (hunk 2 above)
Comment (topeogunleye) [reply queries.sql]: Hi, @momo-87,

You’re very welcome. I’m glad to hear that you made the necessary changes to your project. You have done a great job of following the instructions and applying the SQL concepts. Your code is well-written, clear, and efficient.

File: queries.sql (hunk 4 above)
Comment (topeogunleye) [reply queries.sql:93]: Hi, @momo-87,

You’re very welcome. I’m glad to hear that you made the necessary changes to your project. You have done a great job of following the instructions and applying the SQL concepts. Your code is well-written, clear, and efficient.

Edit: Thank you for your reply. I apologize for the confusion, yes, there is no error in the code.

File: queries.sql (hunk 3 above)
Comment (topeogunleye) [reply queries.sql]: Hi, @momo-87,

You’re very welcome. I’m glad to hear that you made the necessary changes to your project. You have done a great job of following the instructions and applying the SQL concepts. Your code is well-written, clear, and efficient.

**Pull request merged**